pkg/utils/utils_virtual.go

```diff
@@ -0,0 +1,312 @@
+		return
+	}
+
+	if err = json.Unmarshal(rawBytes, networkData); err != nil {
+		glog.Errorf("error unmarshalling raw bytes %v from %s", err, ospNetworkData)
+		return
+	}
+
+	metaData = &OSPMetaData{}
+
+	rawBytes, err = ioutil.ReadFile(ospMetaData)
+	if err != nil {
+		glog.Errorf("error reading file %s, %v", ospMetaData, err)
+		return
+	}
+
+	if err = json.Unmarshal(rawBytes, metaData); err != nil {
+		glog.Errorf("error unmarshalling raw bytes %v from %s", err, ospNetworkData)
+		return
+	}
+
+	return
+}
+
+func parseMetaData(pciAddr string, metaData *OSPMetaData, networkData *OSPNetworkData) (networkID string, networkTag string) {
+
+	if metaData == nil || networkData == nil {
+		return
+	}
+
+	for _, device := range metaData.Devices {
+		if pciAddr == device.Address {
+			for _, link := range networkData.Links {
+				if device.Mac == link.EthernetMac {
+					for _, network := range networkData.Networks {
+						if network.Link == link.ID {
+							networkID = network.NetworkID
+							networkTag = device.Tags
+						}
+					}
+				}
+			}
+		}
+	}
+
+	return
+}
+
+// DiscoverSriovDevicesVirtual discovers VFs on a virtual platform
+func DiscoverSriovDevicesVirtual() ([]sriovnetworkv1.InterfaceExt, error) {
+	glog.V(2).Info("DiscoverSriovDevicesVirtual")
+	pfList := []sriovnetworkv1.InterfaceExt{}
+
+	pci, err := ghw.PCI()
+	if err != nil {
+		return nil, fmt.Errorf("DiscoverSriovDevicesVirtual(): error getting PCI info: %v", err)
+	}
+
+	devices := pci.ListDevices()
+	if len(devices) == 0 {
+		return nil, fmt.Errorf("DiscoverSriovDevicesVirtual(): could not retrieve PCI devices")
+	}
+
+	for _, device := range devices {
+		devClass, err := strconv.ParseInt(device.Class.ID, 16, 64)
+		if err != nil {
+			glog.Warningf("DiscoverSriovDevicesVirtual(): unable to parse device class for device %+v %q", device, err)
+			continue
+		}
+		if devClass != netClass {
+			// Not network device
+			continue
+		}
+
+		if isVirtio(device.Address) {
+			// Do not include virtio devices
+			continue
+		}
+		metaData, networkData := readMetaData()
+		networkID, networkTag := parseMetaData(device.Address, metaData, networkData)
+
+		driver, err := dputils.GetDriverName(device.Address)
+		if err != nil {
+			glog.Warningf("DiscoverSriovDevicesVirtual(): unable to parse device driver for device %+v %q", device, err)
+			continue
+		}
+		iface := sriovnetworkv1.InterfaceExt{
+			PciAddress: device.Address,
+			Driver:     driver,
+			Vendor:     device.Vendor.ID,
+			DeviceID:   device.Product.ID,
+			NetworkID:  networkID,
+			NetworkTag: networkTag,
+		}
+		if mtu := getNetdevMTU(device.Address); mtu > 0 {
+			iface.Mtu = mtu
+		}
+		if name := tryGetInterfaceName(device.Address); name != "" {
+			iface.Name = name
+			iface.Mac = getNetDevMac(name)
+			iface.LinkSpeed = getNetDevLinkSpeed(name)
+		}
+		iface.LinkType = getLinkType(iface)
+
+		iface.TotalVfs = 1
+		iface.NumVfs = 1
+
+		vf := sriovnetworkv1.VirtualFunction{
```

Comment:
I think it would be good to treat VFs as individual devices.  The problem is that the current code seems to be keyed to TotalVFs/NumVfs and I took the approach of setting to 1 to reduce the number code changes.  I can take a look at your suggestion.